### PR TITLE
shardddl: add lock and version for optimistic info

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 
- second ddl's lock may be deleted when first ddl resolved.
- new info putted to etcd may be deleted when previous ddl resolved

### What is changed and how it works?
- add lock in handleInfo and handleOperation to avoid concurrency conflict
- record the version of info, only delete info if all version of info are greater or equal then that in etcd.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by lance6716** [comment]
LGTM, maybe we could test this behaviour in this PR of before GA

File: tests/shardddl3/run.sh [hunk 1]
```diff
@@ -479,6 +479,55 @@ function DM_103() {
...
+    run_sql_source2 "alter table ${shardddl1}.${tb1} drop column b;"
+    run_sql_source2 "alter table ${shardddl1}.${tb1} change a a bigint default 10;"
+    run_sql_source2 "alter table ${shardddl1}.${tb2} drop column b;"
+    run_sql_source2 "alter table ${shardddl1}.${tb2} change a a bigint default 10;"
+
+    check_sync_diff $WORK_DIR $cur/conf/diff_config.toml
+}
+
+function DM_RemoveLock() {
+    ps aux | grep dm-master |awk '{print $2}'|xargs kill || true
```
Comment (lance6716) [inline tests/shardddl3/run.sh:506]: seems we could use `killall dm-master` next time. it's OK here

**Review by lance6716 — COMMENTED** [review]
LGTM again

File: dm/master/shardddl/optimist.go [hunk 2]
```diff
@@ -500,8 +516,7 @@ func (o *Optimist) handleLock(info optimism.Info, tts []optimism.TargetTable, sk
 
 	lock := o.lk.FindLock(lockID)
 	if lock == nil {
-		// this aways means others remove the lock concurrently when resolved ddl.
-		// simply try again.
+		// should not happen
```
Comment (csuzhangxc) [inline dm/master/shardddl/optimist.go:519]: can we remove `return o.handleLock(info, tts, skipDone)` in L 521 now or report an error?

File: dm/master/shardddl/optimist.go [hunk 3]
```diff
@@ -528,35 +543,50 @@ func (o *Optimist) handleLock(info optimism.Info, tts []optimism.TargetTable, sk
...
 	infos := make([]optimism.Info, 0)
 	ops := make([]optimism.Operation, 0)
 	for source, schemaTables := range lock.Ready() {
 		for schema, tables := range schemaTables {
 			for table := range tables {
-				// NOTE: we rely on only `task`, `source`, `upSchema`, and `upTable` used for deletion.
-				infos = append(infos, optimism.NewInfo(lock.Task, source, schema, table, lock.DownSchema, lock.DownTable, nil, nil, nil))
+				// NOTE: we rely on only `task`, `source`, `upSchema`, `upTable` and `Version` used for deletion.
+				info := optimism.NewInfo(lock.Task, source, schema, table, lock.DownSchema, lock.DownTable, nil, nil, nil)
+				info.Version = lock.Versions[source][schema][table]
```
Comment (csuzhangxc) [inline dm/master/shardddl/optimist.go:574]: Do we need `mutex` to protect the read of `lock.Versions`?

File: dm/master/shardddl/optimist.go [hunk 4]
```diff
@@ -528,35 +543,50 @@ func (o *Optimist) handleLock(info optimism.Info, tts []optimism.TargetTable, sk
...
 	if err != nil {
-		return err
+		return deleted, err
 	}
-	o.logger.Info("delete shard DDL infos and lock operations", zap.String("lock", lock.ID), zap.Int64("revision", rev))
-	return nil
+	if deleted {
+		o.logger.Info("delete shard DDL infos and lock operations", zap.String("lock", lock.ID), zap.Int64("revision", rev))
+	} else {
+		o.logger.Info("fail to delete shard DDL infos and lock operations", zap.String("lock", lock.ID), zap.Int64("revision", rev))
```
Comment (csuzhangxc) [inline dm/master/shardddl/optimist.go:589]: How about removing the similar log in L557?

File: pkg/binlog/reader/util.go [hunk 5]
```diff
@@ -62,8 +62,8 @@ func GetGTIDsForPos(ctx context.Context, r Reader, endPos gmysql.Position) (gtid
 		// NOTE: only update endPos/GTIDs for DDL/XID to get an complete transaction.
 		switch ev := e.Event.(type) {
 		case *replication.QueryEvent:
-			parser2, err := event.GetParserForStatusVars(ev.StatusVars)
-			if err != nil {
+			parser2, err2 := event.GetParserForStatusVars(ev.StatusVars)
+			if err2 != nil {
```
Comment (csuzhangxc) [inline pkg/binlog/reader/util.go:66]: log `err2` in L67?

File: pkg/binlog/reader/util.go [hunk 6]
```diff
@@ -74,8 +74,8 @@ func GetGTIDsForPos(ctx context.Context, r Reader, endPos gmysql.Position) (gtid
 					// GTID not enabled, can't get GTIDs for the position.
 					return nil, errors.Errorf("should have a GTIDEvent before the DDL QueryEvent %+v", e.Header)
 				}
-				err = latestGSet.Update(nextGTIDStr)
-				if err != nil {
+				err2 = latestGSet.Update(nextGTIDStr)
+				if err2 != nil {
```
Comment (csuzhangxc) [inline pkg/binlog/reader/util.go:78]: `Annotatef` `err2` in L79?

**Comment by GMHDBJD** [comment]
make `versions` private, change the name to lowercase in lock_test.go

**Comment by csuzhangxc** [comment]
previous failed with `bash: line 1:   508 Killed`

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Pull request merged**